internal/attestation/azure/snp/validator.go

```diff
@@ -36,249 +41,361 @@ type Validator struct {
+			report.Report.IdKeyDigest,
+		)
+	default:
+		return fmt.Errorf(
+			"configured idkeydigests %x don't contain reported idkeydigest %x",
+			v.config.FirmwareSignerConfig.AcceptedKeyDigests,
+			report.Report.IdKeyDigest,
+		)
+	}
+
+	// No IDKeyDigest that was not expected is present.
+	return nil
+}
+
+// azureInstanceInfo contains the necessary information to establish trust in
+// an Azure CVM.
+type azureInstanceInfo struct {
+	// VCEK is the PEM-encoded VCEK certificate for the attestation report.
+	VCEK []byte
+	// CertChain is the PEM-encoded certificate chain for the attestation report.
+	CertChain []byte
+	// AttestationReport is the attestation report from the vTPM (NVRAM) of the CVM.
+	AttestationReport []byte
+	// RuntimeData is the Azure runtime data from the vTPM (NVRAM) of the CVM.
+	RuntimeData []byte
+	// MAAToken is the token of the MAA for the attestation report, used as a fallback
+	// if the IDKeyDigest cannot be verified.
+	MAAToken string
+}
+
+// attestationWithCerts returns a formatted version of the attestation report and its certificates from the instanceInfo.
+// if the VCEK certificate or the certificate chain is not present, the given getter is used to retrieve them
+// from the AMD KDS.
+func (a *azureInstanceInfo) attestationWithCerts(logger attestation.Logger, getter trust.HTTPSGetter) (*spb.Attestation, error) {
+	report, err := abi.ReportToProto(a.AttestationReport)
 	if err != nil {
-		return nil, fmt.Errorf("loading askPEM: %w", err)
+		return nil, fmt.Errorf("converting report to proto: %w", err)
 	}
 
-	if err = ask.CheckSignatureFrom((*x509.Certificate)(&v.config.AMDRootKey)); err != nil {
-		return nil, &askError{err}
+	// Product info as reported through CPUID[EAX=1]
+	sevProduct := abi.DefaultSevProduct()
+	productName := kds.ProductString(sevProduct)
+
+	att := &spb.Attestation{
+		Report:           report,
+		CertificateChain: &spb.CertificateChain{},
+		Product:          sevProduct,
 	}
 
-	if err = vcek.CheckSignatureFrom(ask); err != nil {
-		return nil, &vcekError{err}
+	// If the VCEK certificate is present, parse it and format it.
+	vcek, err := a.parseVCEK()
+	if err != nil {
+		logger.Warnf("Error parsing VCEK: %v", err)
+	}
+	if vcek != nil {
+		att.CertificateChain.VcekCert = vcek.Raw
+	}
+	// Otherwise, retrieve it from AMD KDS.
+	if att.CertificateChain.VcekCert == nil {
+		logger.Infof("VCEK certificate not present, falling back to retrieving it from AMD KDS")
+		vcekURL := kds.VCEKCertURL(productName, report.GetChipId(), kds.TCBVersion(report.GetReportedTcb()))
+		vcek, err := getter.Get(vcekURL)
+		if err != nil {
+			return nil, fmt.Errorf("retrieving VCEK certificate from AMD KDS: %w", err)
+		}
+		att.CertificateChain.VcekCert = vcek
+	}
+
+	// If the certificate chain is present, parse it and format it.
+	ask, ark, certChainParsingErr := a.parseCertChain()
+	if certChainParsingErr != nil {
+		logger.Warnf("Error parsing certificate chain: %v", certChainParsingErr)
```

Comment:
```suggestion
	ask, ark, err := a.parseCertChain()
	if err != nil {
		logger.Warnf("Error parsing certificate chain: %v", err)
```
nit: for consistency